### ISSUE TITLE
Questions about the code

### ISSUE THREAD

**Issue by t1101675** [MANGA-UOFA/LMReward#2]
Hi, I have some questions about the code, could you please provide some help? Thanks!

1. How many GPUs did you use to run `reinforce.sh`? I only succeeded in running the script on one GPU. When I use multiple GPUs, the training process will stuck at the beginning:
```
myname@ubuntu:~/LMReward$ CUDA_VISIBLE_DEVICES=2,3 bash scripts/reinforce.sh
2022-11-21 05:32:09 | Namespace(baseline=True, config_name='/home/myname/LMReward/checkpoints/t5-base/config.json', criterion='cross_entropy', data='/home/myname/LMReward/data/qqp_kaggle/', denom=100.0, entropy=0
.1, fp32=True, init_flat=False, iter_per_update=4, label_smoothing=0, learning_rate=1e-05, length_ratio=2, max_length=256, max_norm=1.0, max_tokens=2048, min_max_length=8, model_name='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', num_training_steps=100000, num_warmup_steps=4000, num_workers=96, optim='adam', repeat=1, reward_clip=1, reward_model='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', save_dir='/home/myname/LMReward/results/qqp_kaggle/reinforce_test', scheduler='constant', seed=0, sent_level=False, softmax=True, suffixes=['src', 'src'], temperature=1.0, tokenizer_name='/home/myname/LMReward/checkpoints/t5-
base/', topk=2, update_per_log=10, update_per_save=1000, update_per_sync=5000, world_size=2)
2022-11-21 05:32:09 | Building tokenizer                
2022-11-21 05:32:10 | Building dataset...
2022-11-21 05:32:19 | #discarded 17 | #preseved 510048 | #total 510065
2022-11-21 05:32:19 | Bachifying data..
2022-11-21 05:32:19 | Building model...
2022-11-21 05:32:32 | GPU-0 process started
2022-11-21 05:32:32 | GPU-1 process started
2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
2022-11-21 05:32:39 | Building trainer...
2022-11-21 05:32:54 | proc started
2022-11-21 05:32:54 | proc started
2022-11-21 05:33:05 | proc started                      
2022-11-21 05:33:06 | proc started                      
2022-11-21 05:33:06 |    original lengths: [4297, 4272]                                                          
2022-11-21 05:33:06 |    reduced to 4272                
2022-11-21 05:33:06 | Training started                  
2022-11-21 05:33:06 | clear queue of size 0             
2022-11-21 05:33:06 | clear queue of size 0             
2022-11-21 05:33:06 | clear queue of size 0             
2022-11-21 05:33:06 | clear queue of size 0
```

2. The reinforcement training is conducted on semi-supervised data. So should I change `DATA` in `reinforce.sh` to the directory that contains the unpaired data?

3.  When I changed the `DATA` directory and used one GPU to run `reinforce.sh`, the trained model at 2K steps begins to output degenerated sentences and the bleu-4 score is around 3.5 (further training will result in even lower scores). Is that normal behavior?
generated sentences when trained for 2K steps:
```
How is s??
How can I make money online?
How???
What is the??
Why is??
What do???
What is??
What does it????
What is the???
How do I??
```

**Comment by yongchanghao** [comment]
Hi. Sorry for the late reply.
1. I use 4GPUs by default. If the program stops here, it usually means there are errors occurred. A typical error is OOM. My trick to track the error message is to use `try...except...` in the sampling process.
2. Yes.
3. This is not a normal behavior. Although it is normal to see mild degeneration after an increase, it is unlikely to degenerate drastically in the beginning. Can you let me know the valid BLEU score of your reward model? Another reason I can think of is the scale of the  reward is different and the preset entropy coefficient makes the entropy term dominate the loss. If this is the case, decreasing the entropy coefficient will  help.

**Comment by t1101675** [comment]
The evaluation outputs (on validation set) of the reward model:

```
14885
14885
BLEU [60.27114009095644, 46.27833456535942, 37.00537293004497, 30.30385584008951]
iBLEU [47.57009451900015, 36.17705441880552, 28.658070714247607, 23.242834872799246]
distinct [5.886527403741175, 28.556784821796832]
length-gen 164868
length-ref 168490
```

I will try decreasing the entropy coefficient. Thanks for your help!

**Comment by frankdarkluo** [comment]
> Hi, I have some questions about the code, could you please provide some help? Thanks!
> 
> 1. How many GPUs did you use to run `reinforce.sh`? I only succeeded in running the script on one GPU. When I use multiple GPUs, the training process will stuck at the beginning:
> 
> ```
> myname@ubuntu:~/LMReward$ CUDA_VISIBLE_DEVICES=2,3 bash scripts/reinforce.sh
> 2022-11-21 05:32:09 | Namespace(baseline=True, config_name='/home/myname/LMReward/checkpoints/t5-base/config.json', criterion='cross_entropy', data='/home/myname/LMReward/data/qqp_kaggle/', denom=100.0, entropy=0
> .1, fp32=True, init_flat=False, iter_per_update=4, label_smoothing=0, learning_rate=1e-05, length_ratio=2, max_length=256, max_norm=1.0, max_tokens=2048, min_max_length=8, model_name='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', num_training_steps=100000, num_warmup_steps=4000, num_workers=96, optim='adam', repeat=1, reward_clip=1, reward_model='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', save_dir='/home/myname/LMReward/results/qqp_kaggle/reinforce_test', scheduler='constant', seed=0, sent_level=False, softmax=True, suffixes=['src', 'src'], temperature=1.0, tokenizer_name='/home/myname/LMReward/checkpoints/t5-
> base/', topk=2, update_per_log=10, update_per_save=1000, update_per_sync=5000, world_size=2)
> 2022-11-21 05:32:09 | Building tokenizer                
> 2022-11-21 05:32:10 | Building dataset...
> 2022-11-21 05:32:19 | #discarded 17 | #preseved 510048 | #total 510065
> 2022-11-21 05:32:19 | Bachifying data..
> 2022-11-21 05:32:19 | Building model...
> 2022-11-21 05:32:32 | GPU-0 process started
> 2022-11-21 05:32:32 | GPU-1 process started
> 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> 2022-11-21 05:32:39 | Building trainer...
> 2022-11-21 05:32:54 | proc started
> 2022-11-21 05:32:54 | proc started
> 2022-11-21 05:33:05 | proc started                      
> 2022-11-21 05:33:06 | proc started                      
> 2022-11-21 05:33:06 |    original lengths: [4297, 4272]                                                          
> 2022-11-21 05:33:06 |    reduced to 4272                
> 2022-11-21 05:33:06 | Training started                  
> 2022-11-21 05:33:06 | clear queue of size 0             
> 2022-11-21 05:33:06 | clear queue of size 0             
> 2022-11-21 05:33:06 | clear queue of size 0             
> 2022-11-21 05:33:06 | clear queue of size 0
> ```

I wonder how you fix this problem? Where is the problem that got the training process got stuck, thanks!

**Comment by t1101675** [comment]
> > Hi, I have some questions about the code, could you please provide some help? Thanks!
> > 
> > 1. How many GPUs did you use to run `reinforce.sh`? I only succeeded in running the script on one GPU. When I use multiple GPUs, the training process will stuck at the beginning:
> > 
> > ```
> > myname@ubuntu:~/LMReward$ CUDA_VISIBLE_DEVICES=2,3 bash scripts/reinforce.sh
> > 2022-11-21 05:32:09 | Namespace(baseline=True, config_name='/home/myname/LMReward/checkpoints/t5-base/config.json', criterion='cross_entropy', data='/home/myname/LMReward/data/qqp_kaggle/', denom=100.0, entropy=0
> > .1, fp32=True, init_flat=False, iter_per_update=4, label_smoothing=0, learning_rate=1e-05, length_ratio=2, max_length=256, max_norm=1.0, max_tokens=2048, min_max_length=8, model_name='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', num_training_steps=100000, num_warmup_steps=4000, num_workers=96, optim='adam', repeat=1, reward_clip=1, reward_model='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', save_dir='/home/myname/LMReward/results/qqp_kaggle/reinforce_test', scheduler='constant', seed=0, sent_level=False, softmax=True, suffixes=['src', 'src'], temperature=1.0, tokenizer_name='/home/myname/LMReward/checkpoints/t5-
> > base/', topk=2, update_per_log=10, update_per_save=1000, update_per_sync=5000, world_size=2)
> > 2022-11-21 05:32:09 | Building tokenizer                
> > 2022-11-21 05:32:10 | Building dataset...
> > 2022-11-21 05:32:19 | #discarded 17 | #preseved 510048 | #total 510065
> > 2022-11-21 05:32:19 | Bachifying data..
> > 2022-11-21 05:32:19 | Building model...
> > 2022-11-21 05:32:32 | GPU-0 process started
> > 2022-11-21 05:32:32 | GPU-1 process started
> > 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> > 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> > 2022-11-21 05:32:39 | Building trainer...
> > 2022-11-21 05:32:54 | proc started
> > 2022-11-21 05:32:54 | proc started
> > 2022-11-21 05:33:05 | proc started                      
> > 2022-11-21 05:33:06 | proc started                      
> > 2022-11-21 05:33:06 |    original lengths: [4297, 4272]                                                          
> > 2022-11-21 05:33:06 |    reduced to 4272                
> > 2022-11-21 05:33:06 | Training started                  
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0
> > ```
> 
> I wonder how you fix this problem? Where is the problem that got the training process got stuck, thanks!

Sorry, I didn't fix this problem. I run the scripts with single GPU in the end.

**Comment by yongchanghao** [comment]
> > Hi, I have some questions about the code, could you please provide some help? Thanks!
> > 
> > 1. How many GPUs did you use to run `reinforce.sh`? I only succeeded in running the script on one GPU. When I use multiple GPUs, the training process will stuck at the beginning:
> > 
> > ```
> > myname@ubuntu:~/LMReward$ CUDA_VISIBLE_DEVICES=2,3 bash scripts/reinforce.sh
> > 2022-11-21 05:32:09 | Namespace(baseline=True, config_name='/home/myname/LMReward/checkpoints/t5-base/config.json', criterion='cross_entropy', data='/home/myname/LMReward/data/qqp_kaggle/', denom=100.0, entropy=0
> > .1, fp32=True, init_flat=False, iter_per_update=4, label_smoothing=0, learning_rate=1e-05, length_ratio=2, max_length=256, max_norm=1.0, max_tokens=2048, min_max_length=8, model_name='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', num_training_steps=100000, num_warmup_steps=4000, num_workers=96, optim='adam', repeat=1, reward_clip=1, reward_model='/home/myname/LMReward/results/qqp_kaggle/reward/model-043k', save_dir='/home/myname/LMReward/results/qqp_kaggle/reinforce_test', scheduler='constant', seed=0, sent_level=False, softmax=True, suffixes=['src', 'src'], temperature=1.0, tokenizer_name='/home/myname/LMReward/checkpoints/t5-
> > base/', topk=2, update_per_log=10, update_per_save=1000, update_per_sync=5000, world_size=2)
> > 2022-11-21 05:32:09 | Building tokenizer                
> > 2022-11-21 05:32:10 | Building dataset...
> > 2022-11-21 05:32:19 | #discarded 17 | #preseved 510048 | #total 510065
> > 2022-11-21 05:32:19 | Bachifying data..
> > 2022-11-21 05:32:19 | Building model...
> > 2022-11-21 05:32:32 | GPU-0 process started
> > 2022-11-21 05:32:32 | GPU-1 process started
> > 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> > 2022-11-21 05:32:34 | loading weight from /home/myname/LMReward/results/qqp_kaggle/reward/model-043k
> > 2022-11-21 05:32:39 | Building trainer...
> > 2022-11-21 05:32:54 | proc started
> > 2022-11-21 05:32:54 | proc started
> > 2022-11-21 05:33:05 | proc started                      
> > 2022-11-21 05:33:06 | proc started                      
> > 2022-11-21 05:33:06 |    original lengths: [4297, 4272]                                                          
> > 2022-11-21 05:33:06 |    reduced to 4272                
> > 2022-11-21 05:33:06 | Training started                  
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0             
> > 2022-11-21 05:33:06 | clear queue of size 0
> > ```
> 
> I wonder how you fix this problem? Where is the problem that got the training process got stuck, thanks!

Sorry I didn't monitor the issue. Are you still having this problem? Is it persistent or appearing occasionally?

**Comment by frankdarkluo** [comment]
I have fixed the problem somehow, by installing the packages in `requirements.txt`.
